### ISSUE TITLE
Is it possible to do a join with relations/foreign keys?

### ISSUE THREAD

**Issue by richardwu** [prisma/prisma#13517]
## Problem

I sometimes find myself wanting to join 2 tables that are related by a column, but don't have referential integrity.

For example, let's say I have 2 tables, `Transaction` and `SmartContract` (think Ethereum).

Every `Transaction` belongs to a `SmartContract`, where `Transaction` has a `smartContractAddress` field. My `Transaction` and `SmartContract` table are backfilled async to each other, so it's possible that the `SmartContract` for a given `Transaction` may not exist yet in the table (hence one cannot create a foreign key constraint).

It's useful to be able to perform joins, hopefully with prisma's typed query methods.

## Suggested solution

I would like to be able to do a `JOIN` query using some variant of the following syntax perhaps:

```js
prisma.smartContract.findMany({
  include: {
    transaction: {
      on: ["address", "smartContractAddress"],
     ...
     }
  }
})
```
which would roughly translate to the SQL
```sql
SELECT * FROM "SmartContract" JOIN "Transaction" ON "SmartContract".address = "Transaction"."smartContractAddress"
```

## Alternatives

I guess one can always write raw queries, but having types + auto-generated selection fields is nice.

## Additional context

.

**Comment by janpio** [comment]
No, that is currently not possible. There are some related issues about "weak" references and similar though, but nothing very highly requested.

**Comment by ShanonJackson** [comment]
Also in this boat, Our problem is we have a table highly optimized for writes, no unique constraints, no foreign keys, low index count.

We want to be able to join to this table "opaquely" without the referential integrity being enforced.

This is similar to https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-integrity except not on the database level but instead of the table level, and a third option where you don't want prisma to enforce it because the whole point is to "join" but not "check", yes this means you can end up in an invalid state (orphaned children) but that's an engineering tradeoff you make.

Its actually the same trade off many databases just force on you, except you just want that on a single table rather than all tables.

**Comment by apatrida** [comment]
It is common that high performance databases do not define foreign keys and drop constraints, sometimes they exist in dev but not prod, sometimes in neither.  Some databases can turn them off while leaving them there, others don't.  You have to allow for the high end cases or this is another ORM or SQL generator that doesn't make the cut in the top end of the performance world. JOOQ is a good example in NOT preventing you from doing anything you want in the end.

You can just say "join to this table you don't expect and all is well" and you can infer the types from the table model provided.

**Comment by apatrida** [comment]
Or, you need a way to extend the model from outside the model, in case it is built by introspection.  

Also, if you have migrations, allow for them to exclude foreign keys even if defined in the model.

Both directions should work (introspection w/adding relationships external to generated code, and migrating without adding relationships to db)

**Comment by CaseyHaralson** [comment]
This would be nice to have.

**Comment by richardwu** [comment]
@janpio is 12 thumbs up highly requested now? 😅 

Out of everything else, this is probably the biggest friction point when using prisma vs a query builder like knex, since the alternative is writing raw SQL :(

**Comment by plmc-changkyu** [comment]
must use $queryRaw...

**Comment by 0tii** [comment]
the company i work at wouldnt be able to use prisma productively given by how often we join unrelated tables. Which is a totally valid use case btw?

**Comment by goldilocked** [comment]
I want to add my voice to requesting this feature. 
I have a query I've been working on where it's impossible to add a direct foreign key. I attempted a workaround using a computed variable, but ran into not being able to use nested items for computed variables. 
My choices are to use post query processing, which ruins the efficiency of my query, or to use rawsql, which adds security concerns to my query and ruins the point of using an ORM in the first place

**Comment by ArnaudTA** [comment]
Same I would like to implement a column `externalId` which refers to a uuid in another table but it can be from multiple tables and I don't wan't to add as many `table_name_id` it would requires

**Comment by CodeSmith32** [comment]
In generalized permission systems, it's common to have a single table that represents access grants for various entities in the database. A row in this type of 'grant' table would join a user or 'principal' entity to any other resource entity in the database, representing a grant that principal has for that resource.

Additionally, this 'grant' table often requires a nullable 'parent' field that optionally points to another row in the 'grant' table. This parent field would be assigned when a user creates an entity since they have a grant that allows creating. The newly created entity would be auto-assigned a set of grants (so the user could interact with his new entity), and each of these grants would point back to the grant that allowed the user to create the entity. This is critical so that, if the parent grant is removed, all child grants can also be removed. For instance, in the below diagram, revoking a user's access from an organization should also revoke access from any departments the user created with that grant, but it should **not** revoke any grants to departments the user was separately granted access to.

<img width="1237" height="737" alt="Image" src="https://github.com/user-attachments/assets/51967952-31dc-4ed3-8405-4e23d87197ea" />

To create this using foreign keys for all relationships get complex very quickly. One would have to start by creating a separate 'grant' table for each entity. But this breaks the ability to have a single table that the parent grant points to. Therefore, a column must be added for each table that could grant the creation of entities of the type the grant table is for.

<img width="671" height="360" alt="Image" src="https://github.com/user-attachments/assets/87811fb3-6da7-4303-a48b-1d372d4a2a35" />

In this diagram, the EntGrant table requires two parent columns because, in this example, the user could be granted the ability to create Entities per Organization or per Department.

-----

What's worse is, if we're forced to use raw SQL for this type of join, we lose a lot of the value of Prisma if we try to create a permission system like this, because then, all 'get' and 'list' queries that require permission checks also require raw SQL.
